### ISSUE TITLE
Implement a Production-ready (Permissive) Dynamic Type

### ISSUE THREAD

**Issue by onelson** [influxdata/flux#4811]
Following on from #4644, #4762

Take a look at the `FIXME`, `TODO`, and `XXX` comments in the POC and resolve them.

Some notable open questions:

- Do we _really need_ to add a `Dynamic()` method to the `Value()` interface? _Footprint of this change would be much smaller without._
- Sizing for flatbuffers: https://github.com/influxdata/flux/blob/7ffb1e69927e686b8cf78d6e3bf6b007afb9ad75/semantic/monotype.go#L918
- Provide impls for stubbed things a la https://github.com/influxdata/flux/blob/7ffb1e69927e686b8cf78d6e3bf6b007afb9ad75/ast/ast.go#L361-L363
- General improvements (for correctness) for things like https://github.com/influxdata/flux/blob/7ffb1e69927e686b8cf78d6e3bf6b007afb9ad75/compiler/compiler.go#L78-L82
- Identify places where `Dynamic` is allowed to be passed but shouldn't. **We want folks to marshal their unknown values asap.**
- Update diagnostics where errors related to Dynamic can arise, try to offer specific guidance.
- Identify other functions (beyond `json.parse`) that should return `[dynamic]`

There are `FIXME` and `TODO` comments littered throughout the POC and each should be visited, considered, and resolved before we can deliver the new type to production.

---

I think additional work could be done after this initial push to get `Dynamic` to production. 

Examples: 
- docs/blog posts 
- adding a builtin to allow array literals to work `dynamic(v: ...)`
- improving null handing (see the comment on nulls here: https://github.com/influxdata/flux/pull/4762#issuecomment-1138935725)

At least until there's a new builtin, for the most part this language feature is not something users will really see. The big difference should be: _things that are errors today should no longer be errors_.

Finally, part of the early discussion around `Dynamic` aspired to clean up code paths where we have to sprinkle runtime type checks around. I'm still not sure how feasible this would be -- might be best to defer that sort of refactor until the proposed compiler refactor.

**Comment by onelson** [comment]
Some questions were raised during a recent grooming meeting. In an effort to nutshell the current state of the POC, I double checked some aspects.

- `array.from` will reject `[dynamic]` directly from `json.parse` (`Error: rows should have been a list of records`).
- `array.map` will accept `[dynamic]`.
- `array.from` will accept `[dynamic]` via `array.map` (sort of weird).
- Record construction allows dynamic values as fields (`null` excluded) without an explicit cast but this is probably not so desirable.

In retrospect, `array.from`'s behavior in the POC seems like my mistake. I think with so much of the rest of the space being so permissive, I should have pushed harder to make it possible to `array.from(rows: json.parse(....))` and have it "just work."

Rationale - this actually "works" in the POC,

```
import "array"
import "experimental/json"
import "experimental/array"

payload = "[{\"a\":  1, \"b\":  \"two\"}]"
data = json.parse(data: bytes(v: payload))
rows = array.map(arr: data, fn: (x) => x) // launder the dynamic "records"
array.from(rows: rows)
```
however it will fail with something like `Error: column size mismatch: [1 2]` if a 2nd object is added to the array with a divergent set of fields.

`json.parse` currently returns type variable `A`. Updating the signature to return `dynamic`, while also making receivers more picky about what they will accept will likely be a big breaking change through stdlib and user scripts.

---

One big question for us is, "how aggressive do we need dynamic values to be marshaled?"

Ultimately, I think we can statically enforce where dynamic is allowed to be or we could push the other way such that it is as permissive as possible.

The POC was mainly concerned about making it possible for `json.parse` to produce values it normally chokes on.

**Comment by onelson** [comment]
Discussed with the team to try and determine next steps for this effort.

The TL;DR might be:

> We'd like to see _another proposal_ before moving forward. Perhaps with the hindsight of the POC we can specify the  mechanics more precisely, with attention to "core values" of the flux language.
>
> "Core values" are TBD but might lean towards static type checks for the "unknowns" and treating dynamic values as things that must be cast before use.
>
> There was also some interest in returning to the problem case `json.parse` to see if we could examine a totally different solution (not involving the addition of a Dynamic type).


Some specific things we _more or less_ agreed on, should this move forward:

- runtime errors related to inferred types and actual types should happen at the point of use, not where values are produced.
- ???


There was also talk about the implications of using type variables vs dynamic in signatures. The POC didn't do a good job of making the difference significant.

Dynamic values currently behave much like type vars and will produce similar compile errors *but shouldn't*. Ex: `filter(fn: (r) => r.a == 1 or r.a == "two")` will fail to type check at compile time so the POC impl was insufficient.


With that said, a rollout approach was discussed WRT the compiler/VM refactor proposal.


**Phase 1**

runtime stays dynamic as it is now
ask dynamic what it is and it answers from inner type
basically a more polished version of the POC

**Phase 2**

typed operators --
no type check for static-known values
keep type checks for dynamic

In IR...
instructions have flavors:
- untyped
- dynamic (possibly the same as untyped)
- typed

**Comment by onelson** [comment]
The team read and discussed [Dynamic Typing in a Statically Typed Language](https://www.st.cs.uni-saarland.de/edu/seminare/2005/advanced-fp/papers/berg.pdf)

Some loose notes from the discussion follow.

---

Casting with a convenient way to handle failure is not something flux currently provides, but could be implemented. Providing a default/fallback to the existing cast functions could fill this gap, or there could be fallible versions of each for this.


Q: How do we funnel values away from `Dynamic` and into static types? How to prevent "bleeding" or "infecting"?
A: We can say `Dynamic` is not valid for _table columns_. If tables require static types, this acts as the boundary.

Alternative approach idea: do something like Typescript's `any` and `unknown` concepts.

- `any` will let you do any sort of expression on the value without any check until runtime.
- `unknown` requires the value to be cast with `as` before you can do anything with it.

If we treat composite types (array, object) sort of like `any`, but only allow traversal operations like member expressions and indexing, then treat leaves of the like `unknown` requiring them to be cast before use. Still need to give this some more thought.

:point_up: the `any`/`unknown` split could just as well be enforced by receivers of the value using just a single `Dynamic` type, right? We just make it so the cast functions can accept `Dynamic` and deny them everywhere else.

**Comment by onelson** [comment]
Closing to pursue #5139 which will try to take this in a direction where we limit exposure to `Dynamic` a bit more instead of letting it "pass" by default.
